### ISSUE TITLE
Adding the board to ICeStudio

### ISSUE THREAD

**Issue by benitoss** [ICE-V-Wireless/ICE-V-Wireless#9]
You can add the board to ICEStudio but the command line to upload the bitstream based in python is not gonna work because you can not configure the port in Apio, it is fixed.
I can solve the problem of using different ports.

I have added to ICEStudio: Colorlight 5A-75B , 5A-75E, i9 , IceSugar pro and ECP5 Evaluation Board.
You can see my pull request in ICEStudio with the changes, if you want to do it yourself but you should be solve the problem of the different ports in your upload python script.

Usually the designers of the boards send me a prototype to accelerate the implementation.

My email is benitoss@gmail.com

**Comment by emeb** [comment]
I'm not sure what the problem is - in both the WiFi and USB versions of the python loading scripts it is possible to specify the board IP address or USB device on the command line. Why can't that be passed in from Apio?

**Comment by benitoss** [comment]
No, the problem is that you cannot pass a variable IP address or variable USB port com in the command line from ICEStudio to Apio to program the board.
The solution for the different COM in windows is to generate a Communication Port autodetection system in the windows binary, in Linux is easier because usually the port is the same always.

**Comment by emeb** [comment]
OK, so the problem is that the invocation approach that icestudio uses to run the python loader scripts doesn't allow you to specify command line options for things like the serial port device or IP address. Your proposal is that the python loader script must autodiscover the appropriate IP address or COM port to use when running in Windows.

For WiFi use: It's *impossible* for the python script to determine what the IP address is - that's assigned by DHCP running on your LAN router and nothing short of an IP scan on your LAN will turn it up. The embedded firmware on the ICE-V-Wireless does support mDNS and the 'send_c3sock.py' script knows the correct mDNS name to use, but that only works if you've set up mDNS on your Windows box and not all users will have done so.

For USB Serial use: Autodetecting the USB CDC COM port that Windows assigns when the device is enumerated is not something I'm well-versed in. It may be possible to use the ESP32C3 VID:PID to cross-reference the COM port and I'll do some research to see if/how to do that.

**Comment by emeb** [comment]
I've added some logic to the `send_c3usb.py` script which autodetects the ESP32C3 USB serial port on Linux. It *should* work on Windows as well, but I have some issues getting python scripts to work properly on Windows so testing by someone with a Windows system and python experience would be helpful.

**Comment by benitoss** [comment]
I can test your python code but I don't have a ESP32C3 USB chip to test. 
Will your python code work with other USB chip ? 
As CH340, CP2102 , FT232H or FT2232H USB chips ??

**Comment by emeb** [comment]
No, these python scripts work exclusively with the ICE-V-Wireless. They use a unique protocol to communicate over WiFi or USB-Serial with the ESP32C3 on the board and will not work with any other ice40 boards that are currently available.

At this time we don't have any more prototype hardware available for beta testing. There may be production boards available at some point in the future after we've fulfilled the obligations to the GroupGets backers, but that will take some time.

**Comment by benitoss** [comment]
Without a prototype is really complicated to generate a compatible driver for ICEStudio. The main idea is working with the USB-Serial connection of the ESP32C3 and work later with the Wi-Fi communication.
Let's wait them for the prototype

**Comment by emeb** [comment]
We're looking a ways to support this. Looks like the best approach will be to add the ICE-V-Wireless to the supported boards list in APIO as described here:

https://apiodoc.readthedocs.io/en/stable/source/contribute/support_new_board.html

For the most part this seems to be a matter of adding to some JSON files, but there will be some additional effort required to provide our Python programming script(s) as packages that can be automatically fetched by the build system when the board is specified at init time. It's worth noting that the docs above do mention that a number of configuration settings can be passed down to the programmer from APIO using the ${variable} syntax, so apparently our serial port or IP socket addresses do not need to be hard-coded or automatically determined.

**Comment by benitoss** [comment]
Yes, that is the way how to add your board to the APIO, but your serial COM port is not allways the same. You need to pass that variable from ICEStudio to APIO and that  variable cannot be passed.
Take the example of Alhambra Board, ICEStudio do not pass any information of the port of the FT2232H programmer included in the board. APIO uses iceprog aplication to program the board using only the VPI and VCI of the FT2232H to autodetect the  COM port.
You have to do something similar and create your aplication program that detects automatically the COM port.
And aditionally, think that ICEStudio works in Linux, Windows and Mac, so the name of the ports are totally different.

**Comment by ojg78** [comment]
Is this work ongoing? It will be disappointing to not be able to use the open source tool chain with my shiny new board.

**Comment by mwelling** [comment]
@ojg78 The open source toolchain suite can and is being used for ICE-V Wireless. This is one of the primary reason we used the ICE40 FPGA.

This ticket is for interfacing the board with the IceStudio graphical front end via APIO. It is on-going.

**Comment by ojg78** [comment]
Understood, thanks for the update and clarifying.
